File: long contest/NAIPC/nwerc2017all/nwerc2017all/glyphrecognition/submissions/accepted/paul-binary-search.cpp
#include <bits/stdc++.h>
using namespace std;

struct point {
	double x, y;
};

const double eps = 1e-9;

int ccw(point a, point b, point c) {
	double cross = (b.x-a.x)*(c.y-b.y) - (c.x-b.x)*(b.y-a.y);
	return abs(cross) > eps ? (cross > 0 ? 1 : -1) : 0;
}

bool point_in_polygon(vector<point> &poly, point a) {
	int winding = 0;
	for (int i = 0; i < poly.size(); i++) {
		point p = poly[i], q = poly[(i+1) % poly.size()];
		
		if (p.y < a.y + eps) {
			if (q.y > a.y + eps && ccw(p,q,a) > 0) winding++;
		} else {
			if (q.y < a.y + eps && ccw(p,q,a) < 0) winding--;
		}
	}
	assert(abs(winding) <= 1);
	return winding;
}

int main() {
	int n; cin >> n;

	vector<point> a(n);
	for (int i = 0; i < n; i++) cin >> a[i].x >> a[i].y;
	
	auto binary_search = [&](int k, bool is_outer) {
		double rmin = 0, rmax = 5e6;
		for (int t = 0; t < 100; t++) {
			double r = (rmin+rmax)/2;
			
			vector<point> poly(k);
			for (int i = 0; i < k; i++) poly[i] = {r*cos(i*2*M_PI/k), r*sin(i*2*M_PI/k)};
			
			int inside_cnt = 0;
			for (int i = 0; i < n; i++) inside_cnt += point_in_polygon(poly,a[i]);
			
			if (is_outer) {
				if (inside_cnt == n) rmax = r; else rmin = r;
			} else {
				if (inside_cnt == 0) rmin = r; else rmax = r;
			}
		}
		return rmin;
	};

	int best_k = -1;
	double best_score = 0;
	for (int k = 3; k <= 8; k++) {
		double rmin = binary_search(k, false), rmax = binary_search(k, true);
		
		double score = (rmin*rmin) / (rmax*rmax);
		if (score > best_score) {
			best_k = k;
			best_score = score;
		}
		cerr << k << " " << score << endl;
	}
	cout << best_k << " " << fixed << setprecision(10) << best_score << endl;
}

File: long contest/NAIPC/nwerc2017all/nwerc2017all/glyphrecognition/submissions/accepted/PaulBinarySearch.java
import java.util.*;
import java.awt.geom.*;

public class PaulBinarySearch {
	
	static double polygonRadius(int k, Point2D p) {
		double rMin = 0, rMax = 5e6;
		for (int t = 0; t < 100; t++) {
			double r = (rMin + rMax) / 2;
			
			Path2D poly = new Path2D.Double();
			poly.moveTo(r,0);
			for (int i = 1; i < k; i++)
				poly.lineTo(r * Math.cos(i*2*Math.PI/k), r * Math.sin(i*2*Math.PI/k));
			poly.closePath();
			
			if (poly.contains(p)) rMax = r; else rMin = r;
		}
		return rMin;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int n = sc.nextInt();
		Point2D[] points = new Point2D[n];
		for (int i = 0; i < n; i++) {
			int x = sc.nextInt(), y = sc.nextInt();
			points[i] = new Point2D.Double(x,y);
		}
		
		int bestK = -1;
		double bestScore = 0;

		for (int k = 3; k <= 8; k++) {
			double rMin = 1e10, rMax = 0;
			for (Point2D p: points) {
				double r = polygonRadius(k, p);
				rMin = Math.min(rMin,r);
				rMax = Math.max(rMax,r);
			}
			double score = (rMin*rMin)/(rMax*rMax);
			if (score > bestScore) {
				bestK = k;
				bestScore = score;
			}
		}
		System.out.println(bestK + " " + bestScore);
	}
}
